From 9736941ea999091ea3bcf45e23f099c9d587b7ee Mon Sep 17 00:00:00 2001
From: Pascal Paillet <p.paillet@foss.st.com>
Date: Wed, 11 Dec 2024 11:26:09 +0100
Subject: [PATCH 1529/1695] drivers: stm32_exti: add an API to read the IMR
 value

Add a function to read the IMR value of a bank.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 724804e11d0e00d646cb2144cb4290514534f59c drivers: stm32_exti: add an API to read the IMR value]
Change-Id: Ic5ce30a204c54c225132ab9522e43480f099ac9a
Signed-off-by: Pascal Paillet <p.paillet@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/429400
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Tested-by: Pascal PAILLET <p.paillet@st.com>
Reviewed-by: Pascal PAILLET <p.paillet@st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/drivers/stm32_exti.c         | 7 +++++++
 core/include/drivers/stm32_exti.h | 2 ++
 2 files changed, 9 insertions(+)

diff --git a/core/drivers/stm32_exti.c b/core/drivers/stm32_exti.c
index 124f78824..e4926e690 100644
--- a/core/drivers/stm32_exti.c
+++ b/core/drivers/stm32_exti.c
@@ -184,6 +184,13 @@ void stm32_exti_unmask(struct stm32_exti_pdata *exti, uint32_t exti_line)
 	cpu_spin_unlock_xrestore(&exti->lock, exceptions);
 }
 
+uint32_t stm32_exti_read_imr(struct stm32_exti_pdata *exti, unsigned int bank)
+{
+	assert(bank < _EXTI_BANK_NR);
+
+	return io_read32(exti->base + _EXTI_C1IMR(bank));
+}
+
 void stm32_exti_enable_wake(struct stm32_exti_pdata *exti, uint32_t exti_line)
 {
 	uint32_t mask = BIT(exti_line % 32);
diff --git a/core/include/drivers/stm32_exti.h b/core/include/drivers/stm32_exti.h
index 70be6fb9a..07a246c45 100644
--- a/core/include/drivers/stm32_exti.h
+++ b/core/include/drivers/stm32_exti.h
@@ -74,4 +74,6 @@ void stm32_exti_set_gpio_port_sel(struct stm32_exti_pdata *exti, uint8_t bank,
  */
 void stm32_exti_set_tz(struct stm32_exti_pdata *exti, uint32_t exti_line);
 
+uint32_t stm32_exti_read_imr(struct stm32_exti_pdata *exti, unsigned int bank);
+
 #endif /*__STM32_EXTI_H*/
-- 
2.47.2

